### ISSUE TITLE
Unable to use `xsd:date`/`xsd:dateTime` field in forms

### ISSUE THREAD

**Issue by kburger** [FAIRDataTeam/FAIRDataPoint-client#148]
Using server `v1.17.2` and client `v1.17.0`. When using a shape that deals with dates or time, the client seems to break down. I've tested in FireFox and Safari, using this example shape:

```ttl
[ sh:path dcterms:issued ;
  sh:datatype xsd:dateTime ; # or xsd:date
  sh:minCount 1 ;
  dash:viewer dash:LiteralViewer ;
  dash:editor dash:DateTimePickerEditor ; # or dash:DatePickerEditor
]
```

When creating new metadata with this shape, the form fails with the following error hint on the input field:
`http://www.w3.org/ns/shacl#DatatypeConstraintComponent`

We see the value in the form widget as expected, e.g. `2024-06-04` for a `xsd:date`/`dash:DatePickerEditor` or `2024-06-04 12:00` for a `xsd:dateTime`/`dash:DateTimePickerEditor`. In the RDF preview however, the value being used is always in the format of `"Tue Jun 04 2024 00:00:00 GMT+0200 (Central European Summer Time)"`.

This formatted value will fail on SHACL validation, which expects an ISO8601 formatted date for `xsd:date` or timestamp for `xsd:dateTime`.

We believe this is a more generic description of the issue raised in 

- https://github.com/FAIRDataTeam/FAIRDataPoint-client/issues/147.

**Comment by dennisvang** [comment]
possibly related to 

- FAIRDataTeam/FAIRDataPoint#504

**Comment by dennisvang** [comment]
How to reproduce:

1. run [minimal compose file v1.17]
2. create a catalog
3. create a dataset with an "Issued" date and/or "Modified" date

The last step fails as follows:

![screenshot failed dataset creation](https://github.com/user-attachments/assets/178e1893-3b8e-40b9-81c7-1cbf07d72ad9)

Confirmed the following, based on [minimal compose file v1.16] and [minimal compose file v1.17]:

|                          | fdp v1.16.2 | fdp v1.17.2 |
|---------------------|------------------|-----------------|
| client v1.16.3 |        pass      |       pass      |
| client v1.17.0 |         fail        |       fail         |

which implies that the error was most likely introduced on the client side in v1.17.0.

[minimal compose file v1.17]: https://gist.github.com/dennisvang/f7fb216cdb4fd36fe814af3185e29a32
[minimal compose file v1.16]: https://gist.github.com/dennisvang/2ff43e8203fbdfd5a25a6facf09b6c0a

**Comment by dennisvang** [comment]
also see diff [v1.16.3...v1.17.0]

[v1.16.3...v1.17.0]: https://github.com/FAIRDataTeam/FAIRDataPoint-client/compare/v1.16.3...v1.17.0

**Comment by dennisvang** [comment]
I've added this to the [v2 milestone], because the issue also arises on the current (e4eee3e) `develop` branch for v2.

[v2 milestone]: https://github.com/FAIRDataTeam/FAIRDataPoint-client/milestone/1

**Comment by dennisvang** [comment]
`POST`ing to the `/dataset` endpoint.

Example of *successful* POST body from *v1.16.3*:

```none
@prefix dcat: <http://www.w3.org/ns/dcat#>.
@prefix dct: <http://purl.org/dc/terms/>.
@prefix foaf: <http://xmlns.com/foaf/0.1/>.
@prefix xsd: <http://www.w3.org/2001/XMLSchema#>.
@prefix loc: <http://localhost/>.
@prefix c: <http://localhost/catalog/>.

loc:new
    a dcat:Dataset, dcat:Resource;
    dct:hasVersion "test";
    dct:isPartOf c:e84203ee-4cf7-41b8-b4ef-9c88dac7d094;
    dct:issued "2025-05-21T22:00:00Z"^^xsd:dateTime;
    dct:publisher [ a foaf:Agent; foaf:name "test" ];
    dct:title "test";
    dcat:theme <http://some.theme>.

```

Example of *failing* POST body from *v1.17.0*:

```none
@prefix dcat: <http://www.w3.org/ns/dcat#>.
@prefix dct: <http://purl.org/dc/terms/>.
@prefix foaf: <http://xmlns.com/foaf/0.1/>.
@prefix language: <http://id.loc.gov/vocabulary/iso639-1/>.
@prefix xsd: <http://www.w3.org/2001/XMLSchema#>.
@prefix loc: <http://localhost/>.
@prefix c: <http://localhost/catalog/>.

loc:new
    a dcat:Dataset, dcat:Resource;
    dct:isPartOf c:daed0173-e2ab-41c4-a953-7b4ab80f5da1;
    dct:issued
        "Thu May 22 2025 00:00:00 GMT+0200 (Central European Summer Time)"^^xsd:dateTime;
    dct:language language:en;
    dct:license <http://purl.org/NET/rdflicense/cc-zero1.0>;
    dct:publisher [ a foaf:Agent; foaf:name "test" ];
    dct:title "test";
    dcat:theme <http://some.theme>;
    dcat:version "test".
```

**Comment by dennisvang** [comment]
Found the culprit: `wrappedValue` (introduced in eda2051)

https://github.com/FAIRDataTeam/FAIRDataPoint-client/blob/4af79dda41b030c7eb3a3fb99489e596fda19d63/src/components/ShaclForm/formData.ts#L120-L126

In the call to `$rdf.quad()` (line 126), the RDF *object*, i.e. the third argument (`wrappedValue`), is converted to an rdflib.js `Literal`.


If the RDF *object* is a JS `Date` (as in v1.16.3), it is converted to a *custom* UTC string by rdflib.js [Literal.fromDate()]:

```ts
  static fromDate(value: Date): Literal {
    if (!(value instanceof Date)) {
      throw new TypeError('Invalid argument to Literal.fromDate()')
    }
    let d2 = function (x) {
      return ('' + (100 + x)).slice(1, 3)
    }
    let date = '' + value.getUTCFullYear() + '-' + d2(value.getUTCMonth() + 1) +
      '-' + d2(value.getUTCDate()) + 'T' + d2(value.getUTCHours()) + ':' +
      d2(value.getUTCMinutes()) + ':' + d2(value.getUTCSeconds()) + 'Z'
    return new Literal(date, null, XSD.dateTime)
```

called from [here]:

```ts
  return Literal.fromValue<any>(value)
```

However, the `wrappedValue` in v1.17.0 is already a `Literal` (of a `Date`), so the `Literal` itself is [returned] before the conversion can take place:

```ts
 if (isTerm(value)) {  // a Node subclass or a Collection
    return value as T
  }
```

### side note

I wonder why rdflib.js uses a custom UTC string format instead of e.g. [Date.toISOString()]...
Perhaps it's because the latter includes milliseconds? See e.g. the [JS date time string format]:

```none
YYYY-MM-DDTHH:mm:ss.sssZ
```
But, then, why? [xsd:dateTime] also allows for milliseconds:

```none
'-'? yyyy '-' mm '-' dd 'T' hh ':' mm ':' ss ('.' s+)? (zzzzzz)?
```

[Literal.toString()]: https://linkeddata.github.io/rdflib.js/doc/classes/Literal.html#toString
[source]: https://github.com/linkeddata/rdflib.js/blob/e9f5dc291d2562f0f965a295ee242f94d0f8d056/src/literal.ts#L118
[Literal.fromDate()]: https://github.com/linkeddata/rdflib.js/blob/c0f0dc5447b3e020e01d420a6f862954b0a42659/src/literal.ts#L132-L142

[source]: https://github.com/linkeddata/rdflib.js/blob/c0f0dc5447b3e020e01d420a6f862954b0a42659/src/literal.ts#L168-L170 
[returned]: https://github.com/linkeddata/rdflib.js/blob/c0f0dc5447b3e020e01d420a6f862954b0a42659/src/collection.ts#L28-L30
[here]: https://github.com/linkeddata/rdflib.js/blob/c0f0dc5447b3e020e01d420a6f862954b0a42659/src/collection.ts#L36
[Date.toISOString()]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toISOString
[JS date time string format]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date#date_time_string_format
[xsd:dateTime]: https://www.w3.org/TR/xmlschema-2/#dateTime

**Comment by dennisvang** [comment]
`$rdf.quad()` calls [Literal.fromValue], which adds datatype information based on the supplied value.

If the supplied value is a JS `Date`, it always sets the datatype to `xsd:dateTime`.

My assumption is that eda2051 was intended to enable *overriding* this auto-detected datatype, so it matches the one specified in `field.datatype`.

A quick fix to restore the valid string representation from v1.16.3 while keeping the "custom" datatype:

```ts
        let wrappedValue = value
        if (hasValue && field && field.datatype) {
          wrappedValue = $rdf.Literal.fromValue(value)  // converts to UTC and valid string represenation
          wrappedValue.datatype.value = field.datatype  // overrides the datatype
          ... // override string representation to match the xsd format corresponding to the datatype?
        }
```

In the case of `xsd:date`, the string representation would also need to be modified to match the [xsd:date] format:

`'-'? yyyy '-' mm '-' dd zzzzzz?`

In addition, I think it would be good to store the UTC-offset for `dateTime`, to preserve the information the user entered, as much as possible. Local time can never be restored from a UTC timestamp without knowing the offset or timezone, whereas it is easy to convert the local time to UTC whenever necessary. This is probably something to be discussed in a separate issue.

[Literal.fromValue]: https://github.com/linkeddata/rdflib.js/blob/c0f0dc5447b3e020e01d420a6f862954b0a42659/src/literal.ts#L167
[xsd:date]: https://www.w3.org/TR/xmlschema-2/#date

**Comment by dennisvang** [comment]
The problem of having to extract the date-only representation from a JS `Date` (which is in UTC) is actually self inflicted.

The custom `DatePickerEditor` extends the `DatePicker` from the [vue2-datepicker] package (3rd-party).
This `DatePicker` actually returns a date `string` with [default format] `YYYY-MM-DD`, which is exactly what we need.
This is the same as the value returned by a native browser date picker (HTML input type date).
However, unfortunately, the custom `DatePickerEditor` converts this `string` to a JS `Date` object:

https://github.com/FAIRDataTeam/FAIRDataPoint-client/blob/2a39c152c394cbe2ddab129cd69c041cac596a7e/src/components/ShaclForm/Editor/DatePickerEditor.vue#L45

As a result we need to jump through hoops to recover the original date `string` from the `Date` object.

So, instead of formatting the `Date` in the `wrappedValue`, I think it would make more sense to restore the *original* `DatePicker` behavior. For example, a quick fix would be as follows:

```ts
this.$emit('input', this.type === 'date' ? val : moment(val).toDate())
```

This way we don't need to worry about UTC conversion issues leading to wrong dates.

However, I even wonder why we don't just always return the `string` value from the `DatePicker`, and convert to JS `Date` only where necessary... That's something for another issue, because it involves more refactoring.

Note `this.type` in the above is determined by the DASH value specified in the schema:

https://github.com/FAIRDataTeam/FAIRDataPoint-client/blob/2a39c152c394cbe2ddab129cd69c041cac596a7e/src/components/ShaclForm/FormInput.vue#L9-L23

[default format]: https://github.com/mengxiong10/vue2-datepicker/blob/bb90ec8500b8863e1025329382a40a2cfb351f93/README.md?plain=1#L120

**Comment by dennisvang** [comment]
Also note that `xsd:date` was not properly supported in v1.16.3 either.

**Comment by dennisvang** [comment]
Re-opening this issue because of the following:

After #183, we can successfully set a date-time value for a resource.

However, if we then edit the same resource and remove that value, making the form field blank, we get another validation error upon saving:

```turtle
[] a sh:ValidationReport;
  sh:conforms false;
  rdf4j:truncated false;
  sh:result [ a sh:ValidationResult;
      sh:focusNode <http://localhost/dataset/19acc034-35b0-4b0e-90df-bdea752c6ab4>;
      rsx:shapesGraph rdf4j:SHACLShapeGraph;
      sh:value "NaN-aN-aNTaN:aN:aNZ"^^xsd:dateTime;
      sh:resultPath dcterms:issued;
      sh:sourceConstraintComponent sh:DatatypeConstraintComponent;
      sh:resultSeverity sh:Violation;
      sh:sourceShape [ a sh:PropertyShape;
          sh:order 20;
          sh:path dcterms:issued;
          sh:datatype xsd:dateTime
        ]
    ] .
```

Apparently the empty form field is now converted to `sh:value "NaN-aN-aNTaN:aN:aNZ"^^xsd:dateTime;`.
